freelist_test.go

```diff
@@ -44,6 +44,131 @@ func TestFreelist_release(t *testing.T) {
 	}
 }
 
+type testRange struct {
+	begin, end txid
+}
+
+type testPage struct {
+	id       pgid
+	n        int
+	allocTxn txid
+	freeTxn  txid
+}
+
+var releaseRangeTests = []struct {
+	title         string
+	pagesIn       []testPage
+	releaseRanges []testRange
+	wantFree      []pgid
+}{
+	{
+		title:         "Single pending in range",
+		pagesIn:       []testPage{testPage{id: 3, n: 1, allocTxn: 100, freeTxn: 200}},
+		releaseRanges: []testRange{testRange{1, 300}},
+		wantFree:      []pgid{3},
+	},
+	{
+		title:         "Single pending with minimum end range",
+		pagesIn:       []testPage{testPage{id: 3, n: 1, allocTxn: 100, freeTxn: 200}},
+		releaseRanges: []testRange{testRange{1, 200}},
+		wantFree:      []pgid{3},
+	},
+	{
+		title:         "Single pending outsize minimum end range",
+		pagesIn:       []testPage{testPage{id: 3, n: 1, allocTxn: 100, freeTxn: 200}},
+		releaseRanges: []testRange{testRange{1, 199}},
+		wantFree:      nil,
+	},
+	{
+		title:         "Single pending with minimum begin range",
+		pagesIn:       []testPage{testPage{id: 3, n: 1, allocTxn: 100, freeTxn: 200}},
+		releaseRanges: []testRange{testRange{100, 300}},
+		wantFree:      []pgid{3},
+	},
+	{
+		title:         "Single pending outside minimum begin range",
+		pagesIn:       []testPage{testPage{id: 3, n: 1, allocTxn: 100, freeTxn: 200}},
+		releaseRanges: []testRange{testRange{101, 300}},
+		wantFree:      nil,
+	},
+	{
+		title:         "Single pending in minimum range",
+		pagesIn:       []testPage{testPage{id: 3, n: 1, allocTxn: 199, freeTxn: 200}},
+		releaseRanges: []testRange{testRange{199, 200}},
+		wantFree:      []pgid{3},
+	},
+	{
+		title:         "Single pending and read transaction at 199",
+		pagesIn:       []testPage{testPage{id: 3, n: 1, allocTxn: 199, freeTxn: 200}},
+		releaseRanges: []testRange{testRange{100, 198}, testRange{200, 300}},
+		wantFree:      nil,
+	},
+	{
+		title: "Adjacent pending and read transactions at 199, 200",
+		pagesIn: []testPage{
+			testPage{id: 3, n: 1, allocTxn: 199, freeTxn: 200},
+			testPage{id: 4, n: 1, allocTxn: 200, freeTxn: 201},
+		},
+		releaseRanges: []testRange{
+			testRange{100, 198},
+			testRange{200, 199},
+			testRange{201, 300},
+		},
+		wantFree: nil,
+	},
+	{
+		title: "Out of order ranges",
+		pagesIn: []testPage{
+			testPage{id: 3, n: 1, allocTxn: 199, freeTxn: 200},
+			testPage{id: 4, n: 1, allocTxn: 200, freeTxn: 201},
+		},
+		releaseRanges: []testRange{
+			testRange{201, 199},
+			testRange{201, 200},
+			testRange{200, 200},
+		},
+		wantFree: nil,
+	},
+	{
+		title: "Multiple pending, read transaction at 150",
+		pagesIn: []testPage{
+			testPage{id: 3, n: 1, allocTxn: 100, freeTxn: 200},
+			testPage{id: 4, n: 1, allocTxn: 100, freeTxn: 125},
+			testPage{id: 5, n: 1, allocTxn: 125, freeTxn: 150},
+			testPage{id: 6, n: 1, allocTxn: 150, freeTxn: 175},
+			testPage{id: 7, n: 1, allocTxn: 175, freeTxn: 200},
```

Comment:
Can we have a testPage where the allocTxn is in the first release range and the freeTxn is in the second releaseRange? (Eg 125, 175)

Reply:
Excellent. Added.